UiTalkAsyncHandler.cs

```diff
@@ -0,0 +1,398 @@
+                  textToTranslate,
+                  ClientState.ClientLanguage.Humanize(),
+                  ClientState.ClientLanguage.Humanize(),
+                  this.configuration.TranslateNpcNames ? this.currentNameTranslation : string.Empty,
+                  this.currentTalkTranslation,
+                  langDict[languageInt].Code,
+                  this.configuration.ChosenTransEngine,
+                  DateTime.Now,
+                  DateTime.Now);
+                string result = InsertTalkData(translatedTalkData);
+                PluginLog.Debug($"Talk Message DB Insert operation result: {result}");
+              }
+
+              return;
+            }
+
+            if (this.configuration.TranslateNpcNames)
+            {
+              this.currentNameTranslationId = Environment.TickCount;
+              this.currentNameTranslation = Resources.WaitingForTranslation;
+              int nameId = this.currentNameTranslationId;
+              string nameTranslation = foundTalkMessage.TranslatedSenderName;
+              this.nameTranslationSemaphore.Wait();
+              if (nameId == this.currentNameTranslationId)
+              {
+                this.currentNameTranslation = nameTranslation;
+              }
+
+              this.nameTranslationSemaphore.Release();
+            }
+
+            this.currentTalkTranslationId = Environment.TickCount;
+            this.currentTalkTranslation = Resources.WaitingForTranslation;
+            int id = this.currentTalkTranslationId;
+            string translatedTalkMessage = foundTalkMessage.TranslatedTalkMessage;
+            this.talkTranslationSemaphore.Wait();
+            if (id == this.currentTalkTranslationId)
+            {
+              this.currentTalkTranslation = translatedTalkMessage;
+            }
+
+            this.talkTranslationSemaphore.Release();
+          }
+          catch (Exception e)
+          {
+            PluginLog.Warning("TranslateTalkUsingImGuiWithoutSwapping Exception: " + e);
+          }
+        });
+    }
+
+    private unsafe void TranslateTalkUsingImGui(string nameToTranslate, string textToTranslate)
+    {
+      PluginLog.Debug($"TranslateTalkUsingImGui: {nameToTranslate}: {textToTranslate}");
+
+      if (this.configuration.SwapTextsUsingImGui)
+      {
+        this.translatedName = string.Empty;
+        this.translatedText = string.Empty;
+        this.TranslateTalkUsingImGuiAndSwapping(nameToTranslate, textToTranslate);
+        return;
+      }
+
+      this.TranslateTalkUsingImGuiWithoutSwapping(nameToTranslate, textToTranslate);
+    }
+
+    private unsafe void UiTalkAsyncHandler(AddonEvent type, AddonArgs args)
+    {
+      PluginLog.Debug($"UiTalkAsyncHandler: {type} {args.AddonName}");
+
+      if (!this.configuration.TranslateTalk)
+      {
+        return;
+      }
+
+      if (args is not AddonRefreshArgs refreshArgs)
+      {
+        this.TranslateTalkReplacing();
+        return;
+      }
```

Comment:
we translate on the refresh event without blocking. We replace on the draw event

Reply:
Nice! I was testing which event sequence was the best yesterday.